src/Momento.Sdk/Internal/ScsTopicClient.cs

```diff
@@ -132,72 +132,125 @@ private async Task<TopicPublishResponse> SendPublish(string cacheName, string to
         {
-            while (await subscription.ResponseStream.MoveNext(cancellationToken))
+            while (!cancellationToken.IsCancellationRequested)
             {
-                var message = subscription.ResponseStream.Current;
+                try
+                {
+                    await _subscription.ResponseStream.MoveNext(cancellationToken);
+                }
+                catch (OperationCanceledException)
+                {
+                    break;
+                }
+                catch (Exception e)
+                {
+                    var sdkException = _exceptionMapper.Convert(e);
+                    if (sdkException.ErrorCode == MomentoErrorCode.CANCELLED_ERROR)
+                    {
+                        break;
+                    }
+                    //TODO: Are there other errors that we should break for?
+
+                    _logger.LogTraceTopicSubscriptionError(_cacheName, _topicName, sdkException);
+
+                    Dispose();
+                    _subscription = Subscribe();
+                    
+                    await Task.Delay(5_000, cancellationToken);
+                    continue;
+                }
+
+                var message = _subscription.ResponseStream.Current;
 
                 switch (message.KindCase)
                 {
                     case _SubscriptionItem.KindOneofCase.Item:
+                        _lastSequenceNumber = message.Item.TopicSequenceNumber;
                         switch (message.Item.Value.KindCase)
                         {
                             case _TopicValue.KindOneofCase.Text:
-                                _logger.LogTraceTopicMessageReceived("text", cacheName, topicName);
+                                _logger.LogTraceTopicMessageReceived("text", _cacheName, _topicName);
                                 return new TopicMessage.Text(message.Item.Value);
                             case _TopicValue.KindOneofCase.Binary:
-                                _logger.LogTraceTopicMessageReceived("binary", cacheName, topicName);
+                                _logger.LogTraceTopicMessageReceived("binary", _cacheName, _topicName);
                                 return new TopicMessage.Binary(message.Item.Value);
                             case _TopicValue.KindOneofCase.None:
                             default:
-                                _logger.LogTraceTopicMessageReceived("unknown", cacheName, topicName);
+                                _logger.LogTraceTopicMessageReceived("unknown", _cacheName, _topicName);
                                 break;
                         }
 
                         break;
                     case _SubscriptionItem.KindOneofCase.Discontinuity:
-                        _logger.LogTraceTopicMessageReceived("discontinuity", cacheName, topicName);
+                        _logger.LogTraceTopicMessageReceived("discontinuity", _cacheName, _topicName);
```

Comment:
in case of discontinuity you should update _lastSequenceNumber to the [`new_topic_sequence`](https://github.com/momentohq/client-protos/blob/main/proto/cachepubsub.proto#L136C10-L136C28) number.

Reply:
Also when doing trace log would it be good to include in trace the value of `Discontinuity` token we got?